gtsam/slam/SmartFactorBase.h

```diff
@@ -207,10 +207,17 @@ class SmartFactorBase: public NonlinearFactor {
     Vector ue = cameras.reprojectionError(point, measured_, Fs, E);
     if (body_P_sensor_ && Fs) {
       const Pose3 sensor_P_body = body_P_sensor_->inverse();
+      size_t camera_dim = size_t(traits<CAMERA>::dimension);
+      size_t pose_dim = size_t(traits<Pose3>::dimension);
+
       for (size_t i = 0; i < Fs->size(); i++) {
-        const Pose3 w_Pose_body = cameras[i].pose() * sensor_P_body;
-        Matrix J(6, 6);
-        const Pose3 world_P_body = w_Pose_body.compose(*body_P_sensor_, J);
+        const Pose3 world_P_body = cameras[i].pose() * sensor_P_body;
+        Matrix J = Matrix::Zero(camera_dim, camera_dim);
+        // Call compose to compute Jacobian for camera extrinsics
+        Matrix H(pose_dim, pose_dim);
+        world_P_body.compose(*body_P_sensor_, H);
+        // Assign extrinsics
+        J.block(0, 0, pose_dim, pose_dim) = H;
```

Comment:
You can just pass the block into the optionalJacobian slot. Amazing, right?

Reply:
I tried doing that but it was complaining about something (which I don't remember right now). I am guessing it will fix itself with static matrices.

Reply:
So as per [this link](https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html), casting an `Eigen::Block` expression to a non-const `Matrix` is not possible.

Quoting the specific line:

> When trying to execute the following code
>```
>MatrixXf C = MatrixXf::Zero(3,6);
>cov(x,y, C.block(0,0,3,3));
>```
> 
> the compiler will fail, because it is not possible to convert the expression returned by MatrixXf::block() into a non-const MatrixXf&. This is the case because the compiler wants to protect you from writing your result to a temporary object.

Reply:
Is the same true for fixed matrices, though?

Reply:
Looks like it. I am having compile time issues otherwise, specifically signature mismatch. This is the best explanation since we're trying to convert a block expression to a non-const matrix as a requirement.

![eigen_error](https://user-images.githubusercontent.com/975964/87159394-490e3080-c28f-11ea-873f-da5afaacbe0e.png)